includes/abstracts/abstract-wc-stripe-payment-gateway.php

```diff
@@ -1055,6 +1055,67 @@ public function generate_create_intent_request( $order, $prepared_source ) {
+	 */
+	public function get_level3_data_from_order( $order ) {
+		// Return an empty array if there's no store postcode (since it's
+		// required for level3 data), or if the version is less than 3.0
+		// (because it's not possible to get a store postcode before WC 3.0).
+		if ( WC_Stripe_Helper::is_wc_lt( '3.0' ) ) {
+			return array();
+		}
+
+		// Get the order items. Don't need their keys, only their values.
+		// Order item IDs are used as keys in the original order items array.
+		$order_items = array_values( $order->get_items() );
+		$currency    = $order->get_currency();
+
+		$stripe_line_items = array_map(function( $item ) use ( $currency ) {
+			$product_id          = $item->get_variation_id()
+				? $item->get_variation_id()
+				: $item->get_product_id();
+			$product_description = substr( $item->get_name(), 0, 26 );
+			$quantity            = $item->get_quantity();
+			$unit_cost           = WC_Stripe_Helper::get_stripe_amount( ( $item->get_subtotal() / $quantity ), $currency );
+			$tax_amount          = WC_Stripe_Helper::get_stripe_amount( $item->get_total_tax(), $currency );
+			$discount_amount     = WC_Stripe_Helper::get_stripe_amount( $item->get_subtotal() - $item->get_total(), $currency );
+
+			return (object) array(
+				'product_code'        => (string) $product_id, // Up to 12 characters that uniquely identify the product.
+				'product_description' => $product_description, // Up to 26 characters long describing the product.
+				'unit_cost'           => $unit_cost, // Cost of the product, in cents, as a non-negative integer.
+				'quantity'            => $quantity, // The number of items of this type sold, as a non-negative integer.
+				'tax_amount'          => $tax_amount, // The amount of tax this item had added to it, in cents, as a non-negative integer.
+				'discount_amount'     => $discount_amount, // The amount an item was discounted—if there was a sale,for example, as a non-negative integer.
+			);
+		}, $order_items);
+
+		$level3_data = array(
+			'merchant_reference'   => $order->get_id(), // An alphanumeric string of up to  characters in length. This unique value is assigned by the merchant to identify the order. Also known as an “Order ID”.
+
+			'shipping_amount'      => WC_Stripe_Helper::get_stripe_amount( $order->get_shipping_total() + $order->get_shipping_tax(), $currency), // The shipping cost, in cents, as a non-negative integer.
+			'line_items'           => $stripe_line_items,
+		);
+
+		// The customer’s U.S. shipping ZIP code.
+		$shipping_address_zip = $order->get_shipping_postcode();
+		if ( $this->is_valid_us_zip_code( $shipping_address_zip ) ) {
+			$level3_data['shipping_address_zip'] = $shipping_address_zip;
+		}
+
+		// The merchant’s U.S. shipping ZIP code.
+		$store_postcode = get_option( 'woocommerce_store_postcode' );
+		if ( $this->is_valid_us_zip_code( $store_postcode ) ) {
+			$level3_data['shipping_from_zip'] = $store_postcode;
+		}
```

Comment:
Is there no error handling we need to do here? What happens if the ZIP codes aren't valid? What happens if one is invalid, and the other one isn't?

Reply:
> What happens if the ZIP codes aren't valid?

AFAICT Stripe does not validate the ZIP codes based on any other factor except format, so the validation in `is_valid_us_zip_code()` method must be sufficient.

> What happens if one is invalid, and the other one isn't?

Both fields are optional. The point here is to provide as much data as possible, the combination does not matter.